wayland-server/src/global.rs

```diff
@@ -29,26 +29,32 @@ impl<I: Resource + 'static, U: Send + Sync + 'static, D: GlobalDispatch<I, U> +
         client_id: ClientId,
         _: GlobalId,
         object_id: ObjectId,
-    ) -> Arc<dyn ObjectData<D>> {
+    ) -> Option<Arc<dyn ObjectData<D>>> {
         let handle = DisplayHandle::from(handle.clone());
         let client = Client::from_id(&handle, client_id).expect("Dead client in bind ?!");
         let resource = <I as Resource>::from_id(&handle, object_id)
             .expect("Wrong object_id in GlobalHandler ?!");
 
         let mut new_data = None;
+        let mut protocol_error = None;
 
         <D as GlobalDispatch<I, U>>::bind(
             data,
             &handle,
             &client,
-            New::wrap(resource),
+            New::wrap(resource.clone()),
             &self.data,
-            &mut DataInit { store: &mut new_data },
+            &mut DataInit { store: &mut new_data, error: &mut protocol_error },
         );
 
         match new_data {
-            Some(data) => data,
-            None => panic!(
+            Some(data) => Some(data),
+            None if protocol_error.is_some() => {
+                let error = protocol_error.unwrap();
+                resource.post_error(error.0, error.1);
+                None
```

Comment:
Couldn't we just have this function set some dummy object_data (invoking `custom_init` internally), rather than adding options all over the API ?

Also, this match can be made cleaner by matching over `(new_data, protocol_error)` rather than using an if and an unwrap.